Modules/_asynciomodule.c

```diff
@@ -2695,10 +2696,13 @@ task_step_impl(TaskObj *task, PyObject *exc)
         }
     }
 
-    if (result == NULL) {
+    if ((gen_status == PYGEN_RETURN) || result == NULL) {
         PyObject *et, *ev, *tb;
 
-        if (_PyGen_FetchStopIterationValue(&o) == 0) {
+        if (result != NULL) {
+            o = result;
+        }
+        if (o || (_PyGen_FetchStopIterationValue(&o) == 0)) {
```

Comment:
Should not `_PyGen_Send_NameTBD()` handle this internally and automatically fetch the StopIteration value?

Reply:
`StopIteration` could still be raised by `send` case